### ISSUE TITLE
Error while running SSD_Mobilenet_v1

### ISSUE THREAD

**Issue by 4rmature** [ARM-software/armnn#306]
I am trying to run ssd_mobilenet_v1_0.75_depth_coco model on arm platform.
I am getting the following error:

TfParseException: Could not find required attribute "data_format" in FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/BatchNorm/FusedBatchNorm

Is fuzed batchnorm layer supported for inference by armnn?
Can i get any help in solving this error?
Thanks

**Comment by arovir01** [comment]
FusedBatchNorm is supported by the TfParser. However, it seems to be expecting the attribute "data_format" to be explicitly set, instead of allowing it to default to NHWC. The parser is too restrictive in this regard.

**Comment by arovir01** [comment]
Please see if this patch gets you any further:
https://review.mlplatform.org/#/c/ml/armnn/+/2367/

I have removed the requirement for "data_format" to be explicitly specified in accordance with the TensorFlow specification:
https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/fused-batch-norm

**Comment by 4rmature** [comment]
When i use the master version along with this patch, i get errors while building ArmNN. If i manually make changes in the above source files, i get errors while running UnitTest.

**Comment by 4rmature** [comment]
I get errors using the new master version with above changes while building(cross compiling) armnn with Tensorflow support.
However if I make the changes in the source files from my earlier version and run my model, I get the same error about data_format.

**Comment by arovir01** [comment]
You can try cherry-picking this commit on top of the stable branch you were previously using.

**Comment by 4rmature** [comment]
Yes i was using 19/08, I tried cherry picking as well

It still shows the error

terminate called after throwing an instance of 'armnn::ParseException'
  what():  Could not find required attribute data_format in node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/BatchNorm/FusedBatchNorm  at function ReadMandatoryNodeAttributeImpl [/home/aditya.c/armnn-tf/armnn/src/armnnTfParser/TfParser.cpp:77]

**Comment by arovir01** [comment]
I suspect the cherry-pick was not successful, otherwise `ParseFusedBatchNorm()` would no longer call `ReadMandatoryNodeAttributeImpl()` for 'data_format' (only for 'epsilon'). This is what reading 'data_format' should look like inside `TfParser::ParseFusedBatchNorm()`:

`const std::string dataFormat = ReadOptionalNodeStringAttribute(nodeDef, "data_format", "NHWC");`

**Comment by 4rmature** [comment]
Yes I have rebuilt TfParser with the changes of the cherry pick.
I am not sure why but i am getting the same error.

**Comment by arovir01** [comment]
Try doing a clean build -- just in case. And make sure TfParser.cpp has actually been updated and you are not building the old code. If this doesn't work, I'm not sure what can be wrong.

**Comment by 4rmature** [comment]
I built it from scratch and verified the source files. I will keep looking as to why it is occurring.

**Comment by arovir01** [comment]
Sorry, I'm out of ideas.

**Comment by 4rmature** [comment]
> I suspect the cherry-pick was not successful, otherwise `ParseFusedBatchNorm()` would no longer call `ReadMandatoryNodeAttributeImpl()` for 'data_format' (only for 'epsilon'). This is what reading 'data_format' should look like inside `TfParser::ParseFusedBatchNorm()` after applying the patch:
> 
> `const std::string dataFormat = ReadOptionalNodeStringAttribute(nodeDef, "data_format", "NHWC");`
> 
> PS: With the risk of playing Captain Obvious -- have you rebuilt the TfParser after you did the cherry-pick?

Are there any other instances in the source code where data_format should default to NHWC? 
I have checked and rechecked and have no idea why the error still reappears. 
If this doesn't work, can you explain how to explicitly set data_format as "NHWC" in my application code?

**Comment by arovir01** [comment]
You will need to set the `data_format` attribute on the TF model itself, in order for it to parse. Can you do this and, if you can, how? -- I don't know.

On the other hand, what we need to figure out is why the parsing fails, when it shouldn't -- and only tweak the model as a last resort.

Have you tried using the 19.11 version of the TfParser?

**Comment by 4rmature** [comment]
> You will need to set the `data_format` attribute on the TF model itself, in order for it to parse. Can you do this and, if you can, how? -- I don't know.
> 
> On the other hand, what we need to figure out is why the parsing fails, when it shouldn't -- and only tweak the model as a last resort.
> 
> Have you tried using the 19.11 version of the TfParser?

The 19.11 version does not contain the changes of the cherry pick. It still calls ReadMandatoryNodeImpl function

I’m using the master version, here the changes have been implemented but data_format still does not default to NHWC.

**Comment by 4rmature** [comment]
I have visualized the model to be run (ssd_mobilenet_v1_0.75_depth_coco) on tensorboard and I see the data_format in the node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/BatchNorm/FusedBatchNorm is set to “NHWC”
Is the TfParser failing?

I was able to run ssd_mobilenet_v1_coco_2018_01_28 model which does not contain FusedBatchNorm on ArmNN successfully.

**Comment by arovir01** [comment]
> data_format still does not default to NHWC even though the ReadOptionalNodeStringAttribute() function is called

What does it default to then? Can you perhaps share the model you are trying to parse and then I could debug it -- unless, of course, it is confidential.

**Comment by 4rmature** [comment]
The model is part of the Tensorflow Object Detection model zoo
http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_0.75_depth_300x300_coco14_sync_2018_07_03.tar.gz
Here is the model I want to run for inference

**Comment by 4rmature** [comment]
Sorry for the trouble
But any idea as to why the parser is failing?

**Comment by arovir01** [comment]
I tried parsing the model you linked, but it didn't even get to the `FusedBatchNorm` nodes. The parsing failed much earlier, throwing an expection due to the input type not being supported:
`terminate called after throwing an instance of 'armnn::ParseException'
  what():  Currently only FLOAT and INT32 are supported for tensorflow nodes (apart from Const). Got DT_UINT8 for Node image_tensor  at function LoadNodeDef [/work/devenv_q/armnn/src/armnnTfParser/TfParser.cpp:3212]`

**Comment by 4rmature** [comment]
> I tried parsing the model you linked, but it didn't even get to the `FusedBatchNorm` nodes. The parsing failed much earlier, throwing an exception due to the input type (UINT8) not being supported:
> `terminate called after throwing an instance of 'armnn::ParseException' what(): Currently only FLOAT and INT32 are supported for tensorflow nodes (apart from Const). Got DT_UINT8 for Node image_tensor at function LoadNodeDef`

Okay so I used Tensorflow Graph Transforms tool to remove cycles and remove certain nodes. I kept only nodes required for inference. My input node is Preprocessor/sub which is of type float32. I dont have the image_tensor node which is of type UINT8. 

Here is the model i am using after graph transforms
[model.zip.zip](https://github.com/ARM-software/armnn/files/3981867/model.zip.zip)

**Comment by arovir01** [comment]
Thanks! The error you came across is thrown because the `BiadAdd` layers do not have a  `data_format` attribute as expected by the parser. On the other hand, even after making `data_format` optional and defaulting to NHWC, the parsing still fails because `StridedSlice` is unsupported.

**Comment by 4rmature** [comment]
Oh okay. Sorry for the weeks of trouble haha. Thank you so much for helping me
